### ISSUE TITLE
Helvetica broken with fontant

### ISSUE THREAD

**Issue by syncrostone** [mwshinn/CanD#9]
Helvetica worked for me in the long ago version I used (https://github.com/mwshinn/canvas/blob/548eeb2470fb75213b16f7df13dd46a760a23c9a/canvas.py) but no longer works for me in this version with fontant — I get this error. I’m not sure what I should do to fix this — it would be nice if this defaulted back to subpar behavior with a warning when this happened rather than refusing to run:

> Warning, italic font not found
Warning, bold font not found
Warning, bold-italic font not found
Traceback (most recent call last):
  File "gp_intro_fig.py", line 63, in <module>
    main()
  File "gp_intro_fig.py", line 60, in main
    make_figure()
  File "gp_intro_fig.py", line 51, in make_figure
    c.add_figure_labels([(chr(ord('A')+i), plot_names[i]) ])
  File "/Users/jasminestone/anaconda3/envs/tf1py36/lib/python3.6/site-packages/paranoid/decorators.py", line 126, in _decorated
    returnvalue = func(*args, **kwargs)
  File "/Users/jasminestone/anaconda3/envs/tf1py36/lib/python3.6/site-packages/CanD-0.0.1-py3.6.egg/cand/canvas.py", line 648, in add_figure_labels
  File "/Users/jasminestone/anaconda3/envs/tf1py36/lib/python3.6/site-packages/CanD-0.0.1-py3.6.egg/cand/canvas.py", line 545, in add_text
  File "/Users/jasminestone/anaconda3/envs/tf1py36/lib/python3.6/site-packages/CanD-0.0.1-py3.6.egg/cand/canvas.py", line 181, in _get_font
  File "/Users/jasminestone/anaconda3/envs/tf1py36/lib/python3.6/site-packages/CanD-0.0.1-py3.6.egg/cand/fontant.py", line 491, in find_font
  File "/Users/jasminestone/anaconda3/envs/tf1py36/lib/python3.6/site-packages/CanD-0.0.1-py3.6.egg/cand/fontant.py", line 487, in filter_if_exists
cand.fontant.NoFontFoundError: No selected fonts have weight = "bold", options are:
    "regular"

**Comment by mwshinn** [comment]
It looks to me like you have multiple versions of Helvetica installed.  CanD will use the exact font you specified if you give the exact name, but if you give a partial name, it will show options.  What happens when you replace "Helvetica" with "Helvet"?

(By the way, I highly recommend Nimbus Sans as a good free version of Helvetica.  It's hard to tell the difference between the two, unlike Helvetica and Arial. )

**Comment by syncrostone** [comment]
I get the following options, but when I try specifying each of them, I get the same error as above. I'm guessing figure labels with fontant now requires a real bold version of the font rather than just awkwardly thickening the lines like matplotlib automatically does.

'{'name': 'Helvet'}
Traceback (most recent call last):
  File "gp_intro_fig.py", line 109, in <module>
    make_figure()
  File "gp_intro_fig.py", line 90, in make_figure
    c.set_font("Helvet", size=8, ticksize=6, titlesize=10)
  File "/Users/jasminestone/anaconda3/envs/tf1py36/lib/python3.6/site-packages/CanD-0.0.1-py3.6.egg/cand/canvas.py", line 131, in set_font
  File "/Users/jasminestone/anaconda3/envs/tf1py36/lib/python3.6/site-packages/CanD-0.0.1-py3.6.egg/cand/canvas.py", line 168, in _get_font
  File "/Users/jasminestone/anaconda3/envs/tf1py36/lib/python3.6/site-packages/CanD-0.0.1-py3.6.egg/cand/fontant.py", line 556, in find_font_family
cand.fontant.MultipleFontsFoundError: Font specified in call to Canvas.set_font() was not specific enough.
Please be more specific in specifying font family.
Specify one of the following font names:
    ".Helvetica Neue DeskInterface", "Helvetica", "Helvetica Neue"
'

**Comment by mwshinn** [comment]
The following command will generate a "fontinfo.debug" file.  Could you please post this file?  (If you don't like posting paths from your computer then you can email it to me privately.)

`python3 -c 'import cand; print("\n".join([str((p,cand.fontant.loadttf(p))) for p in cand.fontant.get_fonts()]))' > fontinfo.debug`

> I'm guessing figure labels with fontant now requires a real bold version of the font rather than just awkwardly thickening the lines like matplotlib automatically does.

Do you have a source for matplotlib just thickening the lines?  To my knowledge, matplotlib doesn't thicken the lines, it just silently chooses a different font which is the "best match" to the one you specified, or a random match if there are several "best matches".  (E.g. see https://github.com/matplotlib/matplotlib/blob/11258c27be3a83832bff8659bd1a03ea2ad35aa6/lib/matplotlib/font_manager.py#L1248 )  Matplotlib also supports afm, which CanD does not.

Also, do you know for sure that you don't have the bold version of any of these three fonts installed?

**Comment by syncrostone** [comment]
Just sent, I don't have a source for matplotlib thickening the lines, that was my naive interpretation of what my happen from " For example, sometimes it will randomly substitute stylistic alternatives of fonts you have selected, when multiple fonts match." in the documentation. I didn't think it through much or do much research on it, just trying to think why this might have worked before but not now.

I'm not sure how to check definitively what fonts I have installed and don't remember how I got helvetica installed in the first place unfortunately. I'm also not sure how to check which afm fonts I have. I may try to investigate in a month or so but don't have time to right now

**Comment by mwshinn** [comment]
Okay, I looked into this and have a few insights.  You have two truetype collection ("ttc") versions of Helvetica installed, which includes multiple fonts in one file (unlike "otf" and "ttf", which include one per file).  Both CanD and matplotlib only support accessing the first font in a ttc file.  Matplotlib (and the previous version of CanD) may be substituting a different font for the bold and italic versions of Helvetica.  I updated the FAQ and just pushed a commit (ab8e1fd4d36dd58eac88983b2d08291c09c1ce6b) which address this issue.  Unfortunately I don't think there's anything else we can do here, so I'll close the issue.
